src/black/handle_ipynb_magics.py

```diff
@@ -0,0 +1,388 @@
+        if (
+            isinstance(node.value, ast.Call)
+            and _is_ipython_magic(node.value.func)
+            and isinstance(node.value.func, ast.Attribute)
+            and node.value.func.attr == "run_cell_magic"
+        ):
+            args = []
+            for arg in node.value.args:
+                assert isinstance(arg, ast.Str)
+                args.append(arg.s)
+            header: Optional[str] = f"%%{args[0]}"
+            if args[1]:
+                assert header is not None
+                header += f" {args[1]}"
+            self.header = header
+            self.body = args[2]
+        self.generic_visit(node)
+
+
+class MagicFinder(ast.NodeVisitor):
+    """Visit cell to look for get_ipython calls.
+
+    Note that the source of the abstract syntax tree
+    will already have been processed by IPython's
+    TransformerManager().transform_cell.
+
+    For example,
+
+        %matplotlib inline
+
+    would have been transformed to
+
+        get_ipython().run_line_magic('matplotlib', 'inline')
+
+    and we look for instances of the latter (and likewise for other
+    types of magics).
+    """
+
+    def __init__(self) -> None:
+        """Record where magics occur."""
+        self.magics: Dict[int, List[Tuple[int, str]]] = collections.defaultdict(list)
+
+    def visit_Assign(self, node: ast.Assign) -> None:
+        """Look for system assign magics.
+
+        For example,
+
+            black_version = !black --version
+
+        would have been transformed to
+
+            black_version = get_ipython().getoutput('black --version')
+
+        and we look for instances of the latter.
+        """
+        if (
+            isinstance(node.value, ast.Call)
+            and _is_ipython_magic(node.value.func)
+            and isinstance(node.value.func, ast.Attribute)
+            and node.value.func.attr == "getoutput"
+        ):
+            args = []
+            for arg in node.value.args:
+                assert isinstance(arg, ast.Str)
+                args.append(arg.s)
+            assert args
+            src = f"!{args[0]}"
+            self.magics[node.value.lineno].append(
+                (
+                    node.value.col_offset,
+                    src,
+                )
+            )
+        self.generic_visit(node)
+
+    def visit_Expr(self, node: ast.Expr) -> None:
+        """Look for magics in body of cell.
+
+        For examples,
+
+            !ls
+            !!ls
+            ?ls
+            ??ls
+
+        would (respectively) get transformed to
+
+            get_ipython().system('ls')
+            get_ipython().getoutput('ls')
+            get_ipython().run_line_magic('pinfo', 'ls')
+            get_ipython().run_line_magic('pinfo2', 'ls')
+
+        and we look for instances of any of the latter.
+        """
+        if isinstance(node.value, ast.Call) and _is_ipython_magic(node.value.func):
+            assert isinstance(node.value.func, ast.Attribute)  # help mypy
```

Comment:
Using a PEP 647 TypeGuard should make this unnecessary. Mypy doesn't support that very well yet though, so no worries if you're not interested.

Reply:
That's awesome, thanks!

Would this require typing_extensions as a runtime dependency before Python 3.10?

Reply:
We already require typing_extensions for `Final` < Python 3.8 so don't worry :)

Reply:
Cool, done - TypeGuard's really neat, it's been worth it trying to contribute to `black` even if just to learn about TypeGuard's existence!